Cleanup transfers scheduler and handle errors gracefully

IvanBelyakoff opened the pull request:
Handle errors in findBlocksCommand and findNewBlocksCommand gracefully.
Add ErrorCounter type for async package and a FiniteCommandWithErrorCounter to exit on errors overflow
Add tests

Closes [#11074](https://github.com/status-im/status-desktop/issues/11074)

IvanBelyakoff committed:
fix(wallet): handle errors in findBlocksCommand and
findNewBlocksCommand gracefully.
Add ErrorCounter type for async package.
Add tests

Closes https://github.com/status-im/status-desktop/issues/11074

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (41)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 0ebf23cc | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/1/) | 2024-01-17 12:07:23 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/1//consoleText) |
| :heavy_check_mark: | 0ebf23cc | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/1/) | 2024-01-17 12:08:08 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240117-120438-0ebf23-pr4581.zip) |
| :heavy_check_mark: | 0ebf23cc | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/1/) | 2024-01-17 12:10:10 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240117-120438-0ebf23-pr4581.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 736c18a0 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/2/) | 2024-01-17 12:28:27 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/2//consoleText) |
| :heavy_check_mark: | 736c18a0 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/2/) | 2024-01-17 12:29:02 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240117-122718-736c18-pr4581.zip) |
| :heavy_check_mark: | 736c18a0 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/2/) | 2024-01-17 12:29:39 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240117-122718-736c18-pr4581.aar) |
| | | | | | | |
| :heavy_check_mark: | f71aa0a8 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/3/) | 2024-01-17 14:16:50 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240117-141452-f71aa0-pr4581.zip) |
| :heavy_multiplication_x: | f71aa0a8 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/3/) | 2024-01-17 14:18:07 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/3//consoleText) |
| :heavy_check_mark: | f71aa0a8 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/3/) | 2024-01-17 14:20:18 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240117-141448-f71aa0-pr4581.aar) |
| | | | | | | |
| :heavy_check_mark: | 96c38253 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/4/) | 2024-01-19 12:51:20 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240119-124858-96c382-pr4581.aar) |
| :heavy_multiplication_x: | 96c38253 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/4/) | 2024-01-19 12:51:58 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/4//consoleText) |
| :heavy_check_mark: | 96c38253 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/4/) | 2024-01-19 12:52:42 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240119-124858-96c382-pr4581.zip) |
| | | | | | | |
| :heavy_check_mark: | 16e22bb5 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/5/) | 2024-01-22 11:20:39 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240122-111853-16e22b-pr4581.zip) |
| :heavy_check_mark: | 16e22bb5 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/5/) | 2024-01-22 11:21:18 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240122-111849-16e22b-pr4581.aar) |
| :heavy_multiplication_x: | 16e22bb5 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/5/) | 2024-01-22 11:55:37 | ~36 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/5//consoleText) |
| | | | | | | |
| :x: | 12229f65 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/6/) | 2024-01-22 12:14:37 | ~17 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/6/consoleText) |
| :x: | 12229f65 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/6/) | 2024-01-22 12:14:38 | ~22 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/6/consoleText) |
| :heavy_multiplication_x: | 12229f65 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/6/) | 2024-01-22 12:15:27 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/6//consoleText) |
| :x: | 12229f65 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/7/) | 2024-01-22 12:16:32 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/7/consoleText) |
| | | | | | | |
| :heavy_check_mark: | 15034de5 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/8/) | 2024-01-22 12:30:15 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240122-122835-15034d-pr4581.zip) |
| :heavy_check_mark: | 15034de5 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/7/) | 2024-01-22 12:31:21 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240122-122835-15034d-pr4581.aar) |
| :heavy_check_mark: | 15034de5 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/7/) | 2024-01-22 13:05:12 | ~36 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/7//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 67560122 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/9/) | 2024-01-22 14:03:18 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240122-140136-675601-pr4581.zip) |
| :heavy_check_mark: | 67560122 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/8/) | 2024-01-22 14:03:53 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240122-140132-675601-pr4581.aar) |
| :heavy_multiplication_x: | 67560122 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/8/) | 2024-01-22 14:34:10 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/8//consoleText) |
| :heavy_check_mark: | 67560122 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/10/) | 2024-01-22 17:43:27 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/10//consoleText) |
| | | | | | | |
| :heavy_check_mark: | fdb74826 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/9/) | 2024-01-23 07:22:35 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240123-072058-fdb748-pr4581.aar) |
| :heavy_check_mark: | fdb74826 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/10/) | 2024-01-23 07:23:20 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240123-072058-fdb748-pr4581.zip) |
| :heavy_check_mark: | fdb74826 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/11/) | 2024-01-23 07:57:00 | ~36 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/11//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 79551e36 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/11/) | 2024-01-23 12:12:12 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240123-121054-79551e-pr4581.zip) |
| :heavy_check_mark: | 79551e36 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/10/) | 2024-01-23 12:17:09 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240123-121054-79551e-pr4581.aar) |
| :heavy_check_mark: | 79551e36 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/12/) | 2024-01-23 12:46:26 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/12//consoleText) |
| | | | | | | |
| :heavy_check_mark: | c8e55669 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/12/) | 2024-01-24 14:05:54 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240124-140400-c8e556-pr4581.zip) |
| :heavy_check_mark: | c8e55669 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/11/) | 2024-01-24 14:07:12 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240124-140356-c8e556-pr4581.aar) |
| :heavy_check_mark: | c8e55669 | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/13/) | 2024-01-24 14:40:44 | ~36 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/13//consoleText) |
| | | | | | | |
| :heavy_check_mark: | f16b0507 | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/13/) | 2024-01-25 07:20:11 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240125-071829-f16b05-pr4581.zip) |
| :heavy_check_mark: | f16b0507 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/12/) | 2024-01-25 07:20:48 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240125-071829-f16b05-pr4581.aar) |
| :heavy_check_mark: | f16b0507 | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/14/) | 2024-01-25 07:54:21 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/14//consoleText) |
| | | | | | | |
| :heavy_check_mark: | e560669b | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/14/) | 2024-01-25 14:56:56 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240125-145440-e56066-pr4581.zip) |
| :heavy_check_mark: | e560669b | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/13/) | 2024-01-25 14:56:57 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240125-145440-e56066-pr4581.aar) |
| :heavy_check_mark: | e560669b | [#15](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/15/) | 2024-01-25 15:29:59 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/15//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 0477e58f | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/15/) | 2024-01-26 08:28:09 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240126-082622-0477e5-pr4581.zip) |
| :heavy_check_mark: | 0477e58f | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/14/) | 2024-01-26 08:28:46 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240126-082622-0477e5-pr4581.aar) |
| :heavy_check_mark: | 0477e58f | [#16](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/16/) | 2024-01-26 09:01:46 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/16//consoleText) |
| :heavy_check_mark: | 0477e58f | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4581/2/) | 2024-01-26 13:13:45 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240126-130857-0477e5-pr4581.zip) |
| | | | | | | |
| :heavy_check_mark: | c11fc6ac | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4581/3/) | 2024-01-26 15:40:15 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240126-153631-c11fc6-pr4581.zip) |
| :heavy_check_mark: | c11fc6ac | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4581/15/) | 2024-01-26 15:41:24 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240126-153627-c11fc6-pr4581.aar) |
| :heavy_check_mark: | c11fc6ac | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4581/16/) | 2024-01-26 15:41:34 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240126-153631-c11fc6-pr4581.zip) |
| :heavy_check_mark: | c11fc6ac | [#17](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/17/) | 2024-01-26 16:16:58 | ~40 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4581/17//consoleText) |

dlipicar commented on services/wallet/async/async.go:
```diff
@@ -244,3 +266,85 @@ func (d *QueuedAtomicGroup) onFinish() {
+	return ec.maxErrors
+}
+
+type FiniteCommandWithErrorCounter struct {
```
💯 for reusable components!

dlipicar reviewed (approved):
LGTM!

So the cause for high RPC consumption was infinitely retrying the fetches for some transfers that always failed, and now we're stopping after a couple of retries?
If so:
1) What's causing them to fail like that?
2) What do we do about these transfers? Are they somehow re-processed at some point?

IvanBelyakoff commented:
> * What's causing them to fail like that?
> * What do we do about these transfers? Are they somehow re-processed at some point?

Good questions, @dlipicar 
Actually it was not the goal for the PR to fix high RPC consumption, but rather to fix error handling which was not clear and messed up, and fix the commands lifecycles so that they keep working when needed, stopped when requested.

1) From what I saw the main reason was hitting some limit - either requests per second rate (then we keep retrying and reach the requests per day limit) or total requests per day. And I've rechecked this - we still keep trying loading transfers even in case of errors forever (or if an error happens for fetching blocks and reaches maxError limit, then transfer loading will stop as well) with a little delay (5 sec), so I'm going to fix it now and stop after some retries - thanx for the heads up. A good thing would be to add an exponential backoff instead of specific retry interval, but I wont do it in scope of this PR.
2) Well, if we fail to load a transfer after some retries - I think we should stop it and report an error to the UI that we have troubles loading transfers. But then we either should allow user to retry it or retry it after some long interval under the hood (can't think of a good number here, it does not make much sense, because for the user it means they wont see the transfers, but if we already tried some exponential backoff with a delay of at least some minutes - then we are out of luck in terms of good UX). This is also a design decision.
If we fail to load history blocks - it could be because of the same reasons, or because of non-archival node issue (I think we are not handling it any more now as we access archive nodes AFAIK, but we generally should) - then again we should do the same as for transfers.

So now we are not reprocessing the blocks or transfers until restart. WDYT?

IvanBelyakoff committed:
fix(wallet): Made loadBlocksAndTransfers command FiniteCommand to
keep restarting on error non-stop, removed some tests accordingly.
Fixed a flaky test for loadBlocksAndTransfers command.
Added tests for async.AtomicGroup.
Made tranfersCommand FiniteCommandWithErrorCounter to prevent infinite
restart.

stefandunca commented on services/wallet/async/async_test.go:
```diff
@@ -0,0 +1,59 @@
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-time.After(1 * time.Millisecond): // wait for the first command to fail and terminate the second command
```
This timeout should be bigger than the expected runtime of the fail task and fail the test if triggered before `ctx.Done()`, if I understand the `AtomicGroup` specifications.

IvanBelyakoff replied:
The intention was to fail the first goroutine while the second is spinning and pretending to do something. The timeouts are the same because I think in any execution order (real situations) it should work the same.

IvanBelyakoff replied:
But I will simplify the test, I think it would be better. Thank you for paying attention to it.

stefandunca commented on services/wallet/async/async_test.go:
```diff
@@ -0,0 +1,59 @@
+
+	err := errors.New("error")
+	group.Add(func(ctx context.Context) error {
+		time.Sleep(1 * time.Millisecond)
```
What do you think about removing the sleep (not to make the test take unnecessarily) but adding instead this failure command after the next command instead, ensuring the other is failing if finished after the `1ms` timeout?

IvanBelyakoff replied:
I actually was not sure what happens if I don't have any timeout here and the function returns immediately with error before the second goroutine even started and waits on ctx.Done() - whether ctx.Done case will be executed or not since cancel() for it was already called. Now due to your comments I tested it and it seems to work fine if Done channel is closed before an attempt to read on it. So I will remove the timeout.

stefandunca commented on services/wallet/async/async_test.go:
```diff
@@ -0,0 +1,59 @@
+		}
+	})
+
+	<-group.WaitAsync()
```
What do you think of calling `group.Wait()`? should be enough to prove the point?

stefandunca commented on services/wallet/async/async_test.go:
```diff
@@ -0,0 +1,59 @@
+			case <-ctx.Done():
+				return nil
+			case <-time.After(5 * time.Millisecond):
+				finished = true
```
What do you think of having it more explicitly with a `require.Fail("timeout waiting for the other command")` instead of `finished` variable?

IvanBelyakoff replied:
I think what you suggest is the opposite to the intention of the test.
The description of AtomicGroup says "AtomicGroup terminates as soon as first goroutine terminates.." and that led me to a conclusion that even in case one routine finishes first with no error the group will cancel the rest routines. But the code makes it clear that the rest goroutines will be cancelled only in case of failure. So I wrote a test to avoid the confusion and prevent possible code changes when AtomicGroup will indeed terminate independent of the return value of the first finished goroutine.
So I do want to wait until the second goroutine times out and sets the `finished` flag, otherwise the test will fail if it will finish on ctx.Done() instead.

stefandunca replied:
Thanks for the great clarifications.

> the description of AtomicGroup says "AtomicGroup terminates as soon as first goroutine terminates.." and that led me to a conclusion that even in case one routine finishes first with no error the group will cancel the rest routines

This was my understanding also looking at the original comment, hence the 3ms timeout seemed like an error case.

stefandunca commented on services/wallet/async/async_test.go:
```diff
@@ -0,0 +1,59 @@
+
+	finished := false
+	group.Add(func(ctx context.Context) error {
+		time.Sleep(1 * time.Millisecond)
```
Same as above, how about replacing this sleep with a different order of adding the commands? Because we know the other command needs 5 ms, if this is added after the other we know it has a chance to run without a sleep.

IvanBelyakoff replied:
No, Stefan, I believe this is wrong. The order of adding commands (goroutines in fact) does not define the order of execution - they will be executed in random order. So we can not rely on that, if that is what you mean.

stefandunca replied:
The comment was based on reading the old description :). Thanks for clarifying.

stefandunca reviewed (approved):
Great finding and fix. I'm looking forward to seeing the final solution, indeed there are some challenges as @IvanBelyakoff suggested.
I left some thoughts and questions in the inline comments.

IvanBelyakoff reviewed (commented):
@stefandunca thank you for your feedback. I will make some amendments
